messaging/src/main/java/org/axonframework/messaging/deadletter/InMemoryDeadLetterQueue.java

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) 2010-2022. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.messaging.deadletter;
+
+import org.axonframework.common.AxonThreadFactory;
+import org.axonframework.messaging.Message;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+import static org.axonframework.common.BuilderUtils.assertThat;
+
+/**
+ * In memory implementation of the {@link DeadLetterQueue}.
+ * <p>
+ * Maintains a {@link Deque} per unique {@link QueueIdentifier}. The maximum amount of {@code Deques} contained by this
+ * {@code DeadLetterQueue} is {@code 1024} (configurable through {@link Builder#maxQueues(int)}). The maximum amount of
+ * {@link DeadLetter letters} per queue also defaults to {@code 1024} (configurable through
+ * {@link Builder#maxQueueSize(int)}).
+ *
+ * @param <T> The type of {@link Message} maintained in this {@link DeadLetterQueue}.
+ * @author Steven van Beelen
+ * @since 4.6.0
+ */
+public class InMemoryDeadLetterQueue<T extends Message<?>> extends SchedulingDeadLetterQueue<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    /**
+     * {@link Clock} instance used to set the time on new {@link DeadLetter DeadLetters}. To fix the time while testing
+     * set this value to a constant value.
+     */
+    public static Clock clock = Clock.systemUTC();
```

Comment:
Allowing configuration of the `Clock` like this is used throughout Axon Framework.
However, we could also allow configuration of the `Clock` through the `Builder`.

Thoughts?

Reply:
Maybe would be nice to have a bean providing the clock, and reuse the same in multiple places, instead of having a static one in multiple places? But might be more like an improvement for framework 5?

Reply:
Agreed with you here, @gklijs. It moved to the `GenericDeadLetter` now (meaning it more closely aligns with the `GenericEventMessage` that also has a `Clock`), but I still don't like it.
So, this is definitely something to pick up for AF 5. :+1: